p2p/src/sync/mod.rs

```diff
@@ -89,445 +97,485 @@ where
-                    SyncControlEvent::Connected(peer_id) => {
-                        log::debug!("register peer {peer_id} to sync manager");
-                        let result = self.register_peer(peer_id).await;
-                        self.handle_error(peer_id, result).await?;
-                    }
-                    SyncControlEvent::Disconnected(peer_id) => {
-                        log::debug!("unregister peer {peer_id} from sync manager");
-                        self.unregister_peer(peer_id)
-                    }
-                },
-                block_id = block_rx.recv(), if !self.chainstate_handle.call(|c| c.is_initial_block_download()).await?? => {
-                    let block_id = block_id.ok_or(P2pError::ChannelClosed)?;
+        // Remove the queued block responses associated with the disconnected peer.
+        self.blocks_queue.retain(|(p, _)| p != &peer);
 
-                    match self.chainstate_handle.call(move |this| this.get_block(block_id)).await?? {
-                        Some(block) => {
-                            let _ = self.peer_sync_handle.make_announcement(Announcement::Block(block)).log_err();
-                        }
-                        None => log::error!("CRITICAL: best block not available"),
-                    }
-                }
-            }
+        if self.peers.remove(&peer).is_some() {
+            log::warn!("Unregistering unknown peer: {peer}");
         }
     }
 
-    /// Returns a receiver for the chainstate `NewTip` events.
-    async fn subscribe_to_chainstate_events(
-        &mut self,
-    ) -> crate::Result<mpsc::UnboundedReceiver<Id<Block>>> {
-        let (tx, rx) = mpsc::unbounded_channel();
+    /// Announces the header of a new block to peers.
+    async fn handle_new_tip(&mut self, block_id: Id<Block>) -> Result<()> {
+        let header = self
+            .chainstate_handle
+            .call(move |c| c.get_block(block_id))
+            .await??
+            // This should never happen because this block has just been produced by chainstate.
+            .expect("A new tip block unavailable")
+            .header()
+            .clone();
+        self.messaging_handle.make_announcement(Announcement::Block(header))
+    }
 
-        let subscribe_func =
-            Arc::new(
-                move |chainstate_event: chainstate::ChainstateEvent| match chainstate_event {
-                    chainstate::ChainstateEvent::NewTip(block_id, _) => {
-                        if let Err(e) = tx.send(block_id) {
-                            log::error!("PubSubMessageHandler closed: {e:?}")
-                        }
-                    }
-                },
-            );
+    /// Sends a block to the peer.
+    async fn send_block(&mut self, peer: T::PeerId, block: Id<Block>) -> Result<()> {
+        self.peers
+            .get_mut(&peer)
+            .ok_or(P2pError::PeerError(PeerError::PeerDoesntExist))?
+            .num_blocks_to_send -= 1;
```

Comment:
This is playing with fire. We must ensure somehow that this is checked, and this looks too naked for me. 

Again, I wanna reiterate that there's lots of arithmetic here that needs to be revised for security.

Reply:
Initially I tried to make it harder to break the invariants by making the `PeerContext` fields private, but it was clumsy, so I removed that. I will think more about that.